src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -228,61 +203,325 @@ private static (LazyStringValue Doc, LazyStringValue Name) ExtractDocIdAndName(J
+
+                return result.ChangeVector;
+            }
+        }
+
+        private static int GetDbIdIndex(DocumentsOperationContext context, BlittableJsonReaderObject data, int dbIdIndex)
+        {
+            if (data.TryGet(DbIds, out BlittableJsonReaderArray dbIds))
+            {
+                for (dbIdIndex = 0; dbIdIndex < dbIds.Length; dbIdIndex++)
+                {
+                    if (dbIds[dbIdIndex].Equals(context.Environment.Base64Id) == false)
+                        continue;
+
+                    break;
+                }
+
+                if (dbIdIndex == dbIds.Length)
+                {
+                    dbIds.Modifications = new DynamicJsonArray {context.Environment.Base64Id};
+                }
+            }
+
+            return dbIdIndex;
         }
 
-        public void PutCounter(DocumentsOperationContext context, string documentId, string collection, string name, long value)
+        private void CreateNewCounter(DocumentsOperationContext context, string documentId, string name, int dbIdIndex, long value, long newETag,
+            BlittableJsonReaderObject data)
         {
-            PutCounterImpl(context, documentId, collection, name, null, value);
+            _countersCount++;
+
+            using (GetCounterPartialKey(context, documentId, name, out var counterKey))
+            {
+                RemoveTombstoneIfExists(context, counterKey);
+            }
+
+            using (context.Allocator.Allocate((dbIdIndex + 1) * SizeOfCounterValues, out var newVal))
+            {
+                if (dbIdIndex > 0) 
+                {
+                    Memory.Set(newVal.Ptr, 0, dbIdIndex * SizeOfCounterValues);
+                }
+
+                var newEntry = (CounterValues*)newVal.Ptr + dbIdIndex;
+
+                newEntry->Value = value;
+                newEntry->Etag = newETag;
+                data.Modifications = new DynamicJsonValue(data)
+                {
+                    [name] = new BlittableJsonReaderObject.RawBlob
+                    {
+                        Length = newVal.Length,
+                        Ptr = newVal.Ptr
+                    }
+                };
+            }
         }
 
-        public void PutCounter(DocumentsOperationContext context, string documentId, string collection, string name, string changeVector, long value)
+        private void IncrementExistingCounter(DocumentsOperationContext context, string documentId, string name, long delta, ref bool exists, BlittableJsonReaderObject.RawBlob existingCounter,
+            int dbIdIndex, long newETag, BlittableJsonReaderObject data, long value)
         {
-            PutCounterImpl(context, documentId, collection, name, changeVector, value);
+            var existingCount = existingCounter.Length / SizeOfCounterValues;
+
+            if (dbIdIndex < existingCount)
+            {
+                exists = true;
+                var counter = (CounterValues*)existingCounter.Ptr + dbIdIndex;
```

Comment:
```suggestion
                var counter = &((CounterValues*)existingCounter.Ptr)[dbIdIndex];
```

Is that more readable?